### PR TITLE
Decouple dependency locking from configurations

### PR DESCRIPTION
**Pull request by jvandort**
Dependency locking was written assuming configurations were the only thing that could be resolved. Shortly, this will no longer be true, so we update dependency locking to avoid this assumption.

In practice, this mostly means we introduce the concept of a 'LockId', which is the name attached to the dependencies for a given resolution within a lockfile. For a configuration, the lockId is equal to the configuration name.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/locking/MissingLockStateException.java [hunk 1]
```diff
@@ -27,11 +28,11 @@ public class MissingLockStateException extends RuntimeException implements Resol
...
     static {
         RESOLUTIONS = ImmutableList.of(
-            "To create the lock state, run a task that will resolve that configuration and add '--write-locks' to the command line.",
+            "To create the lock state, run a task that performs dependency resolution and add '--write-locks' to the command line.",
             new DocumentationRegistry().getDocumentationRecommendationFor("information on generating lock state", "dependency_locking", "generating_and_updating_dependency_locks"));
     }
-    public MissingLockStateException(String configurationName) {
-        super("Locking strict mode: Configuration '" + configurationName + "' is locked but does not have lock state.");
+    public MissingLockStateException(DisplayName displayName) {
+        super("Locking strict mode: " + displayName.getCapitalizedDisplayName() + " is locked but does not have lock state.");
```
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/internal/locking/MissingLockStateException.java:35]: ❔ Wouldn't we also want to see the exact lock id that failed to load here?

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/locking/InvalidLockFileException.java [hunk 2]
```diff
@@ -24,12 +24,12 @@
 public class InvalidLockFileException extends RuntimeException implements ResolutionProvider {
     private final String resolutionMessage;
 
-    public InvalidLockFileException(String name, Exception cause, String resolutionMessage) {
-        super("Invalid lock state for " + name, cause);
+    public InvalidLockFileException(String displayName, Exception cause, String resolutionMessage) {
```
Comment (tresat) [inline platforms/software/dependency-management/src/main/java/org/gradle/internal/locking/InvalidLockFileException.java:27]: 💭 Here, too, it's the lock id that matters, that should be in the message as well.

**Review by tresat — APPROVED** [review]

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/locking/MissingLockStateException.java (hunk 1 above)
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/internal/locking/MissingLockStateException.java:35]: Perhaps. This does not change the original message, as displayName.getCapitalizedDisplayName() is equal to `Configuration '<configurationName>'`

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/locking/InvalidLockFileException.java (hunk 2 above)
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/internal/locking/InvalidLockFileException.java:27]: This is a user-facing message, IMO it should use the display name of the thing getting locked -- which should point the user to the thing they want to fix. The lock ID is a unique ID for the thing being resolved, but I would not expect it to always be as pretty for human readable message. 

The difference would be

`Invalid lock state for configuration 'foo'`
vs
`Invalid lock state for 'foo'` 

Unless we want to put both the display name and lock ID in the message

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/ArtifactCollectionCodec.kt [hunk 3]
```diff
@@ -48,7 +48,7 @@ class ArtifactCollectionCodec(
     override suspend fun WriteContext.encode(value: ArtifactCollectionInternal) {
         val visitor = CollectingArtifactVisitor()
         value.visitArtifacts(visitor)
-        writeString(value.resolutionHost.displayName)
+        writeString(value.resolutionHost.displayName().displayName)
```
[the remark this answers is not in the document]
Comment (jvandort) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/ArtifactCollectionCodec.kt]: Agreed this is ugly. I'll add the string returning method back.

File: platforms/software/dependency-management/src/main/java/org/gradle/internal/locking/InvalidLockFileException.java (hunk 2 above)
Comment (tresat) [reply platforms/software/dependency-management/src/main/java/org/gradle/internal/locking/InvalidLockFileException.java:27]: I was imagining both, for when it's no longer true that configurations are the only thing locked.

`Invalid lock state for configuration 'foo' (id: 'foo').` or something similar.  Redundant for now, but better for the future.

**Pull request merged**